### PR TITLE
Fix for fast math `fp_read_unsigned_bin` function to check max size

### PR DESCRIPTION
**Pull request by dgarske**
Fix for fast math `fp_read_unsigned_bin` function to prevent overrunning buffer if input `c` exceeds max size.

### CODE REVIEW

File: wolfcrypt/src/tfm.c [hunk 1]
```diff
@@ -1868,9 +1868,23 @@ void fp_montgomery_reduce(fp_int *a, fp_int *m, fp_digit mp)
 
 void fp_read_unsigned_bin(fp_int *a, const unsigned char *b, int c)
 {
+    /* must leave room for extra digit because of fp_mul_2d below */
```
Comment (toddouska) [inline wolfcrypt/src/tfm.c]: fp_mul_2d doesn't handle max size ints?

File: wolfcrypt/src/tfm.c [hunk 2]
```diff
@@ -1868,9 +1868,23 @@ void fp_montgomery_reduce(fp_int *a, fp_int *m, fp_digit mp)
...
+#if defined(ALT_ECC_SIZE) || defined(HAVE_WOLF_BIGINT)
+  const word32 maxC = (a->size - 1) * sizeof(fp_digit);
```
Comment (toddouska) [inline wolfcrypt/src/tfm.c]: What about the case that doesn't use fp_mul_2d?  That is, why are we shortening the max input in that case?

File: wolfcrypt/src/tfm.c (hunk 1 above)
Comment (dgarske) [inline wolfcrypt/src/tfm.c]: I confirmed multiple ways the overflow occurs when at the FP_SIZE max without the -1, but you make a good point that fp_mul_2d shouldn't allow it. I'll investigate further. Also I'm doing a math lib update so I might find some other things.

File: wolfcrypt/src/tfm.c (hunk 2 above)
Comment (dgarske) [inline wolfcrypt/src/tfm.c]: Good point about the optimize 32-bit case. It should be able to use the full max. I'll fix.

**Comment by dgarske** [comment]
Legit valgrind warning in `fp_read_unsigned_bin` -> `fp_mul_2d`. Passing back to me to investigate further.

**Comment by dgarske** [comment]
retest this please

File: wolfcrypt/src/tfm.c (hunk 1 above)
Comment (dgarske) [inline wolfcrypt/src/tfm.c]: Realized I need to remove this -1 comment. I'm going to rebase -> squash merge -> force push.

**Pull request merged**